fix: state listener observe writes at wrong time (#13516)
cosmos/cosmos-sdk: CHANGELOG.md

* (x/staking) [#12973](https://github.com/cosmos/cosmos-sdk/pull/12973) Removed `stakingkeeper.RandomValidator`. Use `testutil.RandSliceElem(r, sk.GetAllValidators(ctx))` instead.
* (x/gov) [#13160](https://github.com/cosmos/cosmos-sdk/pull/13160) Remove custom marshaling of proposl and voteoption. 
* (types) [#13430](https://github.com/cosmos/cosmos-sdk/pull/13430) Remove unused code `ResponseCheckTx` and `ResponseDeliverTx`

> * (auth) [#13460](https://github.com/cosmos/cosmos-sdk/pull/13460) The `q auth address-by-id` CLI command has been renamed to `q auth address-by-acc-num` to be more explicit. However, the old `address-by-id` version is still kept as an alias, for backwards compatibility.

Reviewer: Wh are there so many unrelated changelog modifications? Was this due to a bad merge/rebase?

Author: It seems some valid changelogs get removed in this commit on main branch: https://github.com/cosmos/cosmos-sdk/commit/9d9f7df3d6e734ae9b806716ab1c4a18d3946237

Reviewer 2: Changelog entry are deleted from unreleased when backported to the latest released version. As effectively it isn't nee change anymore if it exist in older versions. Please do not re-add them.

* (x/auth)[#13780](https://github.com/cosmos/cosmos-sdk/pull/13780) Querying with `id` (type of int64) in `AccountAddressByID` grpc query now throws error, use account-id(type of uint64) instead.
* (snapshots) [14048](https://github.com/cosmos/cosmos-sdk/pull/14048) Move the Snapshot package to the store package. This is done in an effort group all storage related logic under one package. 
* (baseapp) [#14050](https://github.com/cosmos/cosmos-sdk/pull/14050) refactor `ABCIListener` interface to accept go contexts
* (store) [#13516](https://github.com/cosmos/cosmos-sdk/pull/13516) Update State Streaming APIs:

>   * Add method `ListenCommit` to `ABCIListener`
>   * Move `ListeningEnabled` and  `AddListener` methods to `CommitMultiStore`
>   * Remove `CacheWrapWithListeners` from `CacheWrap` and `CacheWrapper` interfaces
>   * Remove listening APIs from the caching layer (it should only listen to the `rootmulti.Store`)
>   * Add three new options to file streaming service constructor.
>   * Propogate the errors returned by `ABCIListeners` methods, if implementation don't want to propogate errors, should return `nil` instead.

Reviewer 3: ```suggestion
  * Add method `ListenCommit` to `ABCIListener`
  * Move `ListeningEnabled` and  `AddListener` methods to `CommitMultiStore`
  * Remove `CacheWrapWithListeners` from `CacheWrap` and `CacheWrapper` interfaces
  * Remove listening APIs from the caching layer (it should only listen to the `rootmulti.Store`)
  * Add three new options to file streaming service constructor.
  * Modify `ABCIListener` such that any error from any methods will halt the app via `panic`
```
